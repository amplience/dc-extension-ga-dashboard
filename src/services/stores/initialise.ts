import {
  contentItemFilter,
  editionFilter,
  slotFilter,
} from '../../stores/ga-query-filters';
import {
  breakdown,
  contentItemIdMapping,
  editionIdMapping,
  gaApiKey,
  gaClientEmail,
  gaClientId,
  gaViewId,
  slotIdMapping,
} from '../../stores/google-analytics';
import { currencyCode, locale } from '../../stores/localization';
import type { ExtensionInstallation } from '../extension-sdk/extension-sdk.service';

export function initialiseStores(config: ExtensionInstallation): void {
  gaViewId.set(config?.googleAnalyticsViewId || '');
  gaClientId.set(config?.googleAnalyticsClientId || '');
  gaClientEmail.set(config?.googleAnalyticsClientEmail);
  gaApiKey.set(config?.googleAnalyticsKey || '');

  contentItemIdMapping.set(config?.mappings?.contentItemId);
  editionIdMapping.set(config?.mappings?.editionId);
  slotIdMapping.set(config?.mappings?.slotId);

  contentItemFilter.set(config?.filters?.contentItemFilter);
  editionFilter.set(config?.filters?.editionFilter);
  slotFilter.set(config?.filters?.slotFilter);

  breakdown.set(
    config?.breakdown || {
      dimension: 'ga:deviceCategory',
      title: 'Device Breakdown',
    }
  );

  if (config?.localization?.locale) {
    locale.set(config?.localization?.locale);
  }

  if (config?.localization?.currencyCode) {
    currencyCode.set(config?.localization?.currencyCode);
  }
}
