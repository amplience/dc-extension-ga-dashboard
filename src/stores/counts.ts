import { writable, derived } from 'svelte/store';
import type {
  SearchIndexUsersCount,
  SearchIndexSearchesCount,
  SearchIndexNoResultsRate,
} from 'dc-management-sdk-js';

export interface Counts {
  usersCount?: SearchIndexUsersCount;
  searchesCount?: SearchIndexSearchesCount;
  noResultsRate?: SearchIndexNoResultsRate;
}

export const counts = writable<Counts>({});

export const searchesPerUser = derived(counts, ($counts) => {
  return $counts?.usersCount?.count && $counts?.searchesCount?.count
    ? Math.round(
        ($counts?.searchesCount?.count / $counts?.usersCount?.count) * 1e1
      ) / 1e1
    : 0;
});
