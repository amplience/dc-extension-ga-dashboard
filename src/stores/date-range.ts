import { formatDateAsISOString } from '../utils/date-format';
import { persistedWritable } from './persisted-writable';

export interface DateRange {
  from: string;
  to: string;
}

export const DAY = 24 * 60 * 60 * 1000;
export const NOW = new Date().setHours(0, 0, 0, 0).valueOf();

export const INITIAL_DATE_RANGE = {
  from: formatDateAsISOString(new Date(NOW - DAY * 7)),
  to: formatDateAsISOString(new Date(NOW - DAY)),
};

export const dateRange = persistedWritable<DateRange>(
  'date-range',
  INITIAL_DATE_RANGE
);
