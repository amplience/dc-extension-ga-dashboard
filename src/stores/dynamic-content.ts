import type { DynamicContent, Hub } from 'dc-management-sdk-js';
import { writable } from 'svelte/store';

export const client = writable<DynamicContent>(null);

export const hub = writable<Hub>(null);
