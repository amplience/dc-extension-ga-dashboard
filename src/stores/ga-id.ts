import { writable as persistedWritable } from './persisted-store';

export const gaId = persistedWritable<string>('ga-id', null);
