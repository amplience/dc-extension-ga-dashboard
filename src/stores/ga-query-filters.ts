import { derived } from 'svelte/store';
import { editionIdMapping } from './google-analytics';
import { selectedEdition } from './selected-edition';
import { writable } from 'svelte/store';

export const contentItemFilter = writable<string>(null);
export const editionFilter = writable<string>(null);
export const slotFilter = writable<string>(null);

export const gaQueryFilter = derived(
  [selectedEdition, editionIdMapping],
  ([$selectedEdition, $editionIdMapping]) => {
    if ($selectedEdition) {
      return $editionIdMapping + '==' + $selectedEdition.id;
    }
    return undefined;
  }
);

export function constructFilter(
  queryFilter: string,
  customFilter: string
): string {
  if (customFilter) {
    return queryFilter ? `${queryFilter};${customFilter}` : customFilter;
  }

  return queryFilter;
}
