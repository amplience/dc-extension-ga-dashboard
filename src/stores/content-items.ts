import { get, writable } from 'svelte/store';
import { client } from './dynamic-content';

function createContentItems() {
  const { subscribe, set, update } = writable([]);

  return {
    subscribe,
    fetch: async (id) => {
      const cachedContentItem = get(contentItems).find(
        (item) => item.id === id
      );
      if (cachedContentItem) {
        return cachedContentItem;
      }
      const contentItem = await get(client).contentItems.get(id);
      update((n) => [...n, ...[contentItem]]);
      return contentItem;
    },
    reset: () => set([]),
  };
}

export const contentItems = createContentItems();
