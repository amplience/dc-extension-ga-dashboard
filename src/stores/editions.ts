import { get, writable } from 'svelte/store';
import { client } from './dynamic-content';

function createEditions() {
  const { subscribe, set, update } = writable([]);

  return {
    subscribe,
    fetch: async (id) => {
      const cachedEdition = get(editions).find((item) => item.id === id);
      if (cachedEdition) {
        return cachedEdition;
      }
      const edition = await get(client).editions.get(id);
      update((n) => [...n, ...[edition]]);
      return edition;
    },
    reset: () => set([]),
  };
}

export const editions = createEditions();
