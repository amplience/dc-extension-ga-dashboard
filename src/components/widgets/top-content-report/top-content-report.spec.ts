import { tick } from 'svelte';
import { render } from '@testing-library/svelte';
import TopContentReport from './top-content-report.svelte';
import {
  getDataReport,
  processReportData,
  RequestTimeout,
} from '../../../stores/gapi';
import { dateRange } from '../../../stores/date-range';

jest.mock('../../../stores/gapi');

describe('TopContentReport', () => {
  beforeEach(() => {
    dateRange.set({ from: '2020-11-01', to: '2020-11-02' });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should render the TopContentReport component', async () => {
    (getDataReport as jest.Mock).mockImplementation(() => ({
      rows: ['dimension', 1, 1, 1, 1],
      totalsForAllResults: {
        'ga:totalEvents': 1,
        'ga:uniqueEvents': 1,
      },
    }));
    (processReportData as jest.Mock).mockImplementation(jest.fn());
    const { container } = render(TopContentReport, {});

    await tick();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should retry getting report data when we get a timeout back', async () => {
    (getDataReport as jest.Mock).mockImplementation(() =>
      Promise.reject(new RequestTimeout('GAPI request timeout'))
    );
    const { container } = render(TopContentReport, {});

    await tick();
    await tick();

    expect((getDataReport as jest.Mock).mock.calls).toMatchSnapshot();
    expect(getDataReport as jest.Mock).toBeCalledTimes(2);
    expect(container.firstChild).toMatchSnapshot();
  });
});
