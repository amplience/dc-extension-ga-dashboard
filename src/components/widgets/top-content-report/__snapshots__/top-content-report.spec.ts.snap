// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopContentReport should render the TopContentReport component 1`] = `
<div>
  <section
    class="top-content-report svelte-126j0ji"
  >
    <article
      class="svelte-pntwh0"
    >
       
      <section
        class="widget-header svelte-1kld2ef"
      >
        <div
          class="title svelte-1kld2ef"
        >
          <h2
            class="svelte-1kld2ef"
          >
            Top content
          </h2>
           
           
        </div>
         
        <div
          class="actions svelte-1kld2ef"
        >
          <div
            slot="actions"
          >
            <div
              class="select top-content-report-sizes-select svelte-c7jdhd"
            >
              <span>
                Show
              </span>
               
              <div
                class="
    mdc-select
    smui-select
    
    
    smui-select--standard
    
    
  "
              >
                 
                <i
                  class="mdc-select__dropdown-icon"
                />
                 
                <input
                  id="SMUI-select-0"
                  type="hidden"
                  value="5"
                />
                 
                <div
                  aria-haspopup="listbox"
                  aria-labelledby="SMUI-select-0-smui-label SMUI-select-0-smui-selected-text"
                  aria-required="false"
                  class="mdc-select__selected-text"
                  id="SMUI-select-0-smui-selected-text"
                  role="button"
                  tabindex="0"
                >
                  5
                </div>
                 
                 
                 
                <div
                  class="
    mdc-line-ripple
    
    
  "
                />
                
                 
              </div>
            </div>
          </div>
        </div>
      </section>
       
      <section
        class="widget-body svelte-4p1s5s"
      >
        <div
          class="spinner-container svelte-wkqaim"
          style="--z-index: 10;"
        >
          <div
            class="spinner dot-0 svelte-wkqaim"
          />
          <div
            class="spinner dot-1 svelte-wkqaim"
          />
          <div
            class="spinner dot-2 svelte-wkqaim"
          />
        </div>
         
        
      </section>
    </article>
  </section>
</div>
`;

exports[`TopContentReport should retry getting report data when we get a timeout back 1`] = `
Array [
  Array [
    null,
    null,
    "5",
    Object {
      "from": "2020-11-01",
      "to": "2020-11-02",
    },
    null,
  ],
  Array [
    null,
    null,
    "5",
    Object {
      "from": "2020-11-01",
      "to": "2020-11-02",
    },
    null,
  ],
]
`;

exports[`TopContentReport should retry getting report data when we get a timeout back 2`] = `
<div>
  <section
    class="top-content-report svelte-126j0ji"
  >
    <article
      class="svelte-pntwh0"
    >
       
      <section
        class="widget-header svelte-1kld2ef"
      >
        <div
          class="title svelte-1kld2ef"
        >
          <h2
            class="svelte-1kld2ef"
          >
            Top content
          </h2>
           
           
        </div>
         
        <div
          class="actions svelte-1kld2ef"
        >
          <div
            slot="actions"
          >
            <div
              class="select top-content-report-sizes-select svelte-c7jdhd"
            >
              <span>
                Show
              </span>
               
              <div
                class="
    mdc-select
    smui-select
    
    
    smui-select--standard
    
    
  "
              >
                 
                <i
                  class="mdc-select__dropdown-icon"
                />
                 
                <input
                  id="SMUI-select-1"
                  type="hidden"
                  value="5"
                />
                 
                <div
                  aria-haspopup="listbox"
                  aria-labelledby="SMUI-select-1-smui-label SMUI-select-1-smui-selected-text"
                  aria-required="false"
                  class="mdc-select__selected-text"
                  id="SMUI-select-1-smui-selected-text"
                  role="button"
                  tabindex="0"
                >
                  5
                </div>
                 
                 
                 
                <div
                  class="
    mdc-line-ripple
    
    
  "
                />
                
                 
              </div>
            </div>
          </div>
        </div>
      </section>
       
      <section
        class="widget-body svelte-4p1s5s"
      >
        <div
          class="spinner-container svelte-wkqaim"
          style="--z-index: 10;"
        >
          <div
            class="spinner dot-0 svelte-wkqaim"
          />
          <div
            class="spinner dot-1 svelte-wkqaim"
          />
          <div
            class="spinner dot-2 svelte-wkqaim"
          />
        </div>
         
        
      </section>
    </article>
  </section>
</div>
`;
